scanpy/preprocessing/normalization.py

```diff
@@ -0,0 +1,204 @@
+import numpy as np
+from scipy.sparse import issparse
+from sklearn.utils import sparsefuncs
+from .. import logging as logg
+
+def _normalize_data(X, counts, after=None, copy=False):
+    X = X.copy() if copy else X
+    after = np.median(counts) if after is None else after
+    counts += (counts == 0)
+    counts /= after
+    if issparse(X):
+        X = sparsefuncs.inplace_row_scale(X, 1/counts)
+    else:
+        X /= counts[:, None]
+    return X if copy else None
+
+def normalize_quantile(data, counts_per_cell_after=None, counts_per_cell=None,
+                       quantile=1, min_counts=1, key_n_counts=None, copy=False,
+                       layers=[], use_rep=None):
+    """Normalize total counts per cell.
+
+    Normalize each cell by total counts over genes, so that every cell has
+    the same total count after normalization.
+
+    Similar functions are used, for example, by Seurat [Satija15]_, Cell Ranger
+    [Zheng17]_ or SPRING [Weinreb17]_.
+
+    Parameters
+    ----------
+    data : :class:`~anndata.AnnData`
+        The annotated data matrix of shape `n_obs` × `n_vars`. Rows correspond
+        to cells and columns to genes.
+    counts_per_cell_after : `float` or `None`, optional (default: `None`)
+        If `None`, after normalization, each cell has a total count equal
+        to the median of the *counts_per_cell* before normalization.
+    counts_per_cell : `np.array`, optional (default: `None`)
+        Precomputed counts per cell.
+    quantile : `float`, optional (default: 1)
+        Only use genes are less than fraction (specified by *quantile*) of the total
+        reads in every cell.
+    min_counts : `int`, optional (default: 1)
+        Cells with counts less than `min_counts` are filtered out during
+        normalization.
+    key_n_counts : `str`, optional (default: `'n_counts'`)
+        Name of the field in `adata.obs` where the total counts per cell are
+        stored.
+    copy : `bool`, optional (default: `False`)
+        If an :class:`~anndata.AnnData` is passed, determines whether a copy
+        is returned.
+    layers : `str` or list of `str`, optional (default: `[]`)
+        List of layers to normalize. Set to `'all'` to normalize all layers.
+    use_rep : `str` or `None`, optional (default: `None`)
```

Comment:
How come that `use_rep` refers layers? It was used in other instances for providing a key to `.obsm`, which might store a compressed representation of the data (e.g. from PCA or an autoencoder). It doesn't make sense at all to have this for keying layers??? Sorry if I missed this previously. 

I cannot stay like this.

Reply:
That might be true. That's why I wrote "sorry if I missed this previously". I don't know whether you or Volker added it, but it's definitely a serious inconsistency to use this parameter name here.   Please, think of a different name.